src/ws/client.rs

```diff
@@ -165,11 +166,39 @@ impl<'a> WebSocketEventType<'a> {
                 }
             }
             esp_websocket_event_id_t_WEBSOCKET_EVENT_CLOSED => Ok(Self::Closed),
+            #[cfg(esp_idf_version_major = "5")]
+            esp_websocket_event_id_t_WEBSOCKET_EVENT_BEFORE_CONNECT => Ok(Self::BeforeConnect),
             _ => Err(EspError::from_infallible::<ESP_ERR_INVALID_ARG>().into()),
         }
     }
 }
 
+/// An encoding of an X.509 certificate or key.
+pub enum EspWebSocketX509Encoding<'a> {
```

Comment:
Look [here](https://github.com/esp-rs/esp-idf-svc/blob/master/src/tls.rs#L72). How about reusing it, rather than coming up with a similar struct yet slightly different?

Reply:
> Look [here](https://github.com/esp-rs/esp-idf-svc/blob/master/src/tls.rs#L72). How about reusing it, rather than coming up with a similar struct yet slightly different?

Thanks for the reference! I'll try using that type.

Reply:
Ok, I've reused the X509 type in 59926e3.

There were some complications.

First I had to extend X509 to track and report its encoding, since EspWebSocketClient must know the encoding to determine what to specify in the esp_websocket_client_config_t fields cert_len, client_cert_len, and client_key_len.

Then I had to generalize _RawCstrs.as_nptr()/as_ptr()_ and _to_cstring_arg()_ to accept an `Into<Vec<u8>>`, since _X509.data()_ returns a `&[u8]`; which also required me to update some _to_cstring_arg()_ callers that were passing `&&str` to pass `&str` instead.

And because _X509.pem()_ takes a CStr, I had to add a null terminator to the SERVER_CERT in the example app.

But because _X509.data()_ includes that null terminator, and _to_cstring_arg()_ creates a CString, whose input cannot contain a null terminator, I had to remove the terminator from the byte slice before passing it to _RawCstrs.as_nptr()_.